batch-processing/pipeline.go

```diff
@@ -0,0 +1,353 @@
+package main
+
+import (
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const (
+	READ     = "read"
+	DOWNLOAD = "download"
+	CONVERT  = "convert"
+	UPLOAD   = "upload"
+	WRITE    = "write"
+)
+
+var TASKS = []string{DOWNLOAD, CONVERT, UPLOAD}
+
+var (
+	OutputHeader       = []string{"url", "input", "output", "s3url", "error"}
+	FailedOutputHeader = []string{"url"}
+)
+
+var (
+	ErrDuplicateURL    = errors.New("duplicate URL, skipping")
+	ErrImageExistsInS3 = errors.New("image already exists in S3, skipping")
+)
+
+// Pipeline struct
+type Pipeline struct {
+	config     *Config
+	workers    map[string]int
+	channels   map[string]chan *Out
+	maxRetries uint64    // Max number of retries for downloading and uploading an image
+	urlMap     *sync.Map // this will record urls that are already processed so we don't download again.
+}
+
+type Task func(wg *sync.WaitGroup)
+
+func (p *Pipeline) Read(csvReader *csv.Reader) {
+	// Read the input CSV file
+	// For each line, send the URL to the download channel
+	defer close(p.channels[READ])
+
+	csvReader.FieldsPerRecord = 1
+
+	for {
+		row, err := csvReader.Read()
+
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			p.channels[READ] <- &Out{Err: err}
+			continue
+		}
+		p.channels[READ] <- &Out{Url: row[0]}
+	}
+	log.Println("Read done")
+}
+
+func (p *Pipeline) Download(wg *sync.WaitGroup) {
+	// Download the image from the URL
+	// Send the image to the convert channel
+	defer wg.Done()
+	var (
+		row  *Out
+		ok   bool
+		body io.Reader
+		hash string
+		ext  string
+		err  error
+	)
+	for {
+
+		row, ok = <-p.channels[READ]
+		if !ok {
+			break
+		}
+		if row.Err != nil {
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		// Check if the URL has already been processed
+		// If it has, skip it
+		_, ok := p.urlMap.Load(row.Url)
+		if ok {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: ErrDuplicateURL,
```

Comment:
Makes sense, thanks! 

I'm filtering them out during read now, which is something I should've done from the beginning rather than complicating things